### PR TITLE
Yet more `parse_tt` improvements

### PR DESCRIPTION
**Pull request by nnethercote**
Including lots of comment improvements, and an overhaul of how `matches` work that gives big speedups.

r? @petrochenkov

### CODE REVIEW

**Comment by nnethercote** [comment]
Some `check full` instruction count results.

Benchmark & Profile | Scenario | % Change | Significance Factor?
-- | -- | -- | --
quote-stress check | full | -17.38% | 86.90x
async-std-1.10.0 check | full | -16.64% | 83.18x
time-macros-0.2.3 check | full | -10.55% | 52.74x
yansi-0.5.0 check | full | -9.62% | 48.10x
deep-vector check | full | -2.85% | 14.26x
num-derive-0.3.3 check | full | -2.51% | 12.56x
pest_generator-2.1.3 check | full | -2.45% | 12.27x
inotify-0.10.0 check | full | -2.30% | 11.49x
scroll_derive-0.11.0 check | full | -2.28% | 11.41x
ctor-0.1.21 check | full | -2.25% | 11.26x
mockall_derive-0.11.0 check | full | -1.98% | 9.90x
futures-macro-0.3.19 check | full | -1.89% | 9.46x
html5ever check | full | -0.85% | 4.27x
token-stream-stress check | full | 0.66% | 3.32x
syn-1.0.89 check | full | -0.48% | 2.38x
hyper-0.14.18 check | full | -0.41% | 2.07x

File: compiler/rustc_expand/src/mbe/macro_parser.rs [hunk 1]
```diff
@@ -254,24 +254,24 @@ crate enum ParseResult<T> {
 /// of metavars to the token trees they bind to.
 crate type NamedParseResult = ParseResult<FxHashMap<MacroRulesNormalizedIdent, NamedMatch>>;
 
-/// Count how many metavars are named in the given matcher `ms`.
-pub(super) fn count_names(ms: &[TokenTree]) -> usize {
-    ms.iter().fold(0, |count, elt| {
-        count
-            + match elt {
-                TokenTree::Delimited(_, delim) => count_names(delim.inner_tts()),
+/// Count how many metavars are named `matcher`.
```
Comment (petrochenkov) [inline compiler/rustc_expand/src/mbe/macro_parser.rs:262]: Looks like some words are missing in this sentence.

File: compiler/rustc_expand/src/mbe/macro_parser.rs [hunk 2]
```diff
@@ -254,24 +254,24 @@ crate enum ParseResult<T> {
...
+pub(super) fn count_metavars(matcher: &[TokenTree]) -> usize {
```
Comment (petrochenkov) [inline compiler/rustc_expand/src/mbe/macro_parser.rs:263]:
```suggestion
pub(super) fn count_metavar_decls(matcher: &[TokenTree]) -> usize {
```

File: compiler/rustc_expand/src/mbe/macro_parser.rs [hunk 3]
```diff
@@ -254,24 +254,24 @@ crate enum ParseResult<T> {
...
                 TokenTree::MetaVarDecl(..) => 1,
-                // Panicking here would abort execution because `parse_tree` makes use of this
-                // function. In other words, RHS meta-variable expressions eventually end-up here.
-                //
-                // `0` is still returned to inform that no meta-variable was found. `Meta-variables
-                // != Meta-variable expressions`
+                // RHS meta-variable expressions eventually end-up here. `0` is returned to inform that
+                // no meta-variable was found, because "meta-variables" != "meta-variable
+                // expressions".
                 TokenTree::MetaVarExpr(..) => 0,
```
Comment (petrochenkov) [inline compiler/rustc_expand/src/mbe/macro_parser.rs:274]: This should be unreachable `count_metavars` should never be called when parsing RHS (`parsing_patterns == false`) in the first place, it's 0 there by definition.

File: compiler/rustc_expand/src/mbe/macro_parser.rs [hunk 4]
```diff
@@ -40,28 +40,28 @@
 //!
 //! Remaining input: a a a b
 //! cur: [a · $( a )* a b]
-//! Descend/Skip (first item).
+//! Descend/Skip (first mp).
```
Comment (petrochenkov) [inline compiler/rustc_expand/src/mbe/macro_parser.rs:43]:
```rust
//! Descend/Skip (first position).
```
I understand using "mp"s in variable names, but in comments I think we can afford being more wordy.

File: compiler/rustc_expand/src/mbe/macro_parser.rs [hunk 5]
```diff
@@ -89,15 +89,13 @@ use std::borrow::Cow;
...
-///
 /// This is used by `parse_tt_inner` to keep track of delimited submatchers that we have
 /// descended into.
 #[derive(Clone)]
-struct MatcherTtFrame<'tt> {
-    /// The "parent" matcher that we are descending into.
-    elts: &'tt [TokenTree],
-    /// The position of the "dot" in `elts` at the time we descended.
+struct MatcherPosFrame<'tt> {
+    /// The "parent" matcher that we have descending from.
```
Comment (petrochenkov) [inline compiler/rustc_expand/src/mbe/macro_parser.rs:96]:
```rust
    /// The "parent" matcher that we have been descending from.
```
Is this a grammatical sentence?

File: compiler/rustc_expand/src/mbe/macro_parser.rs [hunk 6]
```diff
@@ -402,11 +419,21 @@ pub struct TtParser<'tt> {
 
     /// The set of mps that are waiting for the black-box parser.
     bb_mps: Vec<Box<MatcherPos<'tt>>>,
+
+    /// Pre-allocate an empty match array, so it can be cloned cheaply for macros with many rules
+    /// that have no metavars.
+    empty_matches: Lrc<NamedMatchVec>,
```
Comment (petrochenkov) [inline compiler/rustc_expand/src/mbe/macro_parser.rs:425]: Nit: perhaps it's better to move this optimization to a separate commit, the "Overhaul how matches are recorded" is already complex without it.

File: compiler/rustc_expand/src/mbe/macro_parser.rs (hunk 3 above)
Comment (nnethercote) [reply compiler/rustc_expand/src/mbe/macro_parser.rs:274]: I originally thought that too, but if you make it unreachable these tests fail:
```
failures:
    [ui] ui/macros/rfc-3086-metavar-expr/count-and-length-are-distinct.rs
    [ui] ui/macros/rfc-3086-metavar-expr/feature-gate-macro_metavar_expr.rs
    [ui] ui/macros/rfc-3086-metavar-expr/macro-expansion.rs
    [ui] ui/macros/rfc-3086-metavar-expr/out-of-bounds-arguments.rs
```
There may be something to clean up here, but I'd prefer to do it in a follow-up. This change was just meant to give this function a nicer name and slightly cleaner code, it wasn't doing anything about where/how it is called.

File: compiler/rustc_expand/src/mbe/macro_parser.rs (hunk 5 above)
Comment (nnethercote) [reply compiler/rustc_expand/src/mbe/macro_parser.rs:96]: Not really! :)  "we have descended from" is better.

File: compiler/rustc_expand/src/mbe/macro_parser.rs (hunk 6 above)
Comment (nnethercote) [reply compiler/rustc_expand/src/mbe/macro_parser.rs:425]: Ok.

**Comment by nnethercote** [comment]
I have updated the code to address comments. Most of the changes are in the relevant commits, but I split out the `empty_matches` change into a new commit.

File: compiler/rustc_expand/src/mbe/macro_parser.rs (hunk 3 above)
Comment (petrochenkov) [reply compiler/rustc_expand/src/mbe/macro_parser.rs:274]: Ok.
